### PR TITLE
feat: add expo router

### PR DESCRIPTION
**Pull request by PiotrekPKP**
Adds Expo Router as it's already in RC.

### CODE REVIEW

File: apps/expo/tsconfig.json [hunk 1]
```diff
@@ -4,5 +4,5 @@
     "strict": true,
     "jsx": "react-native"
   },
-  "include": ["src", "index.ts", "app.config.ts"]
+  "include": ["app", "utils", "types", "index.ts", "app.config.ts"]
```
Comment (juliusmarminge) [inline apps/expo/tsconfig.json:7]: can we not use src with expo-router 👀 👀

Comment (PiotrekPKP) [reply apps/expo/tsconfig.json:7]: We technically can, but we cannot use src + app dir (expo router looks for app dir to find pages)
That's why I thought that having both app dir and src dir just for types and utils would be kind of weird

Comment (PiotrekPKP) [reply apps/expo/tsconfig.json:7]: Also, I have totally no clue how changing files in expo app makes NextJS's lint fail haha

Comment (juliusmarminge) [reply apps/expo/tsconfig.json:7]: > We technically can, but we cannot use src + app dir (expo router looks for app dir to find pages)
> 
> That's why I thought that having both app dir and src dir just for types and utils would be kind of weird

You should open an issue about that on their part

Comment (PiotrekPKP) [reply apps/expo/tsconfig.json:7]: > > We technically can, but we cannot use src + app dir (expo router looks for app dir to find pages)
> > That's why I thought that having both app dir and src dir just for types and utils would be kind of weird
> 
> You should open an issue about that on their part

Yeah, I definitely will

Comment (PiotrekPKP) [reply apps/expo/tsconfig.json:7]: Issue is open: https://github.com/expo/router/issues/231

**Comment by PiotrekPKP** [comment]
Until the `expo-router` issue is solved, we'd have to use this folder structure. 

Also, I'm not sure why NextJS started having problem with linting and why I had to add `next` dependency to `eslint-config-next`, but it works so... :shrug:

File: packages/config/tailwind/index.js [hunk 2]
```diff
@@ -1,6 +1,12 @@
 /** @type {import('tailwindcss').Config} */
 module.exports = {
-  content: ["./src/**/*.{ts,tsx}", "./src/_app.tsx"],
+  content: [
+    // NextJS
+    "./src/**/*.{ts,tsx}",
+    // Expo
+    "./app/*.{ts,tsx}",
+    "./app/**/*.{ts,tsx}",
+  ],
```
Comment (juliusmarminge) [inline packages/config/tailwind/index.js:9]: Just include `src/**` here then expo Will have to extend the preset and handle "it's special case".

**Review by juliusmarminge — COMMENTED** [review]
I still think the rest of the files should be in src even though expo router doesnt support it (yet)

File: packages/config/tailwind/index.js (hunk 2 above)
Comment (PiotrekPKP) [reply packages/config/tailwind/index.js:9]: Alright, will do it today

Comment (PiotrekPKP) [reply packages/config/tailwind/index.js]: Implemented :)

**Comment by juliusmarminge** [comment]
Question though, is this the proper way to grab params from dynamic paths?

Should we not get something from props? I only get a fixed string `[id]` and not the unique id o the "request"

CleanShot 2023-02-08 at 21 01 33

File: apps/expo/app/index.tsx [hunk 3]
```diff
@@ -1,21 +1,23 @@
...
   onDelete: () => void;
-}> = ({ post, onPress, onDelete }) => {
+}> = ({ post, onDelete }) => {
+  const router = useRouter();
+
   return (
     <View className="flex flex-row rounded-lg bg-white/10 p-4">
       <View className="flex-grow">
-        <TouchableOpacity onPress={onPress}>
+        <TouchableOpacity onPress={() => router.push(`/post/${post.id}`)}>
```
Comment (juliusmarminge) [inline apps/expo/app/index.tsx:18]: would have liked a `<Link href="">` here but that seems to apply some nasty styles which i cba fixing so just used `router.push` instead :/

**Comment by juliusmarminge** [comment]
Changed up some other styles while i was at it:

CleanShot 2023-02-08 at 21 17 31

**Review by juliusmarminge — APPROVED** [review]

**Comment by PiotrekPKP** [comment]
> Question though, is this the proper way to grab params from dynamic paths?
> 
> Should we not get something from props? I only get a fixed string `[id]` and not the unique id o the "request"
> 
> CleanShot 2023-02-08 at 21 01 33

I think we should use `useSearchParams`: https://expo.github.io/router/docs/features/linking#usesearchparams

**Comment by juliusmarminge** [comment]
> 



> > Question though, is this the proper way to grab params from dynamic paths?
> > Should we not get something from props? I only get a fixed string `[id]` and not the unique id o the "request"
> > CleanShot 2023-02-08 at 21 01 33
> 
> I think we should use `useSearchParams`: https://expo.github.io/router/docs/features/linking#usesearchparams

Oh - feel free to change that in a new PR :)

**Comment by PiotrekPKP** [comment]
> > 
> 
> > > Question though, is this the proper way to grab params from dynamic paths?
> > > Should we not get something from props? I only get a fixed string `[id]` and not the unique id o the "request"
> > > CleanShot 2023-02-08 at 21 01 33
> > 
> > 
> > I think we should use `useSearchParams`: https://expo.github.io/router/docs/features/linking#usesearchparams
> 
> Oh - feel free to change that in a new PR :)

Sure, will look at it :)

**Comment by juliusmarminge** [comment]
I still think it should come from props though :/

**Comment by PiotrekPKP** [comment]
> I still think it should come from props though :/

Why? I think the current approach is ok

**Comment by juliusmarminge** [comment]
> Why? I think the current approach is ok

I've just found that this is so much better in Next.js appdir:

```ts
// app/post/[id]/page.tsx
function Page({ params }: { params: { name: string } }) {
  const { data } = getByName(params.name);
}
```
compared to the old:
```ts
// pages/post/[id].tsx
function Page() {
  const { query } = useRouter();
  if (!query.name) throw new Error(); // <-- notice we have to check this
  const { data } = getByName(query.name);
}
```

**Comment by PiotrekPKP** [comment]
> Member
> ###

I honestly haven't seen that option before, thanks for letting me know it exists haha
Anyways, I'm not sure if it's possible to use with Expo Router...

**Pull request merged**